refactor: migrate to maven-index repository for data source

knqyf263 opened the pull request:
## Summary

This PR removes the crawler functionality from trivy-java-db and migrates it to use the maven-index repository as the data source. The architecture is now simplified to focus solely on building SQLite databases from pre-processed TSV files.

### Key Changes:
- Removed crawler package and related dependencies
- Updated build process to use maven-index TSV files  
- Streamlined GitHub Actions workflow
- Added support for forks using dynamic repository owner

## Benefits

- **Simplified architecture**: Focuses on single responsibility of database building
- **Reduced dependencies**: Removed unused crawler-related packages
- **Better separation of concerns**: Crawling logic moved to dedicated repository
- **Fork-friendly**: Workflow automatically adapts to different repository owners
- **Cleaner codebase**: Removed unused functions and packages

## Usage

Build database from maven-index repository:
```bash
make db-build MAVEN_INDEX_DIR=../maven-index
```

The database building process now reads TSV files from the maven-index repository and creates the SQLite database with the same schema and functionality as before.

knqyf263 committed:
refactor: migrate to maven-index repository for data source

This change removes the crawler functionality from trivy-java-db and
migrates it to use the maven-index repository as the data source.

## Overview

- Removed crawler package and related dependencies
- Updated build process to use maven-index TSV files
- Streamlined architecture to focus on database building only
- Updated GitHub Actions workflow to clone maven-index repository

## Changes Made

### Code Changes
- Removed `pkg/crawler` package entirely
- Removed `crawl` command from main.go
- Cleaned up unused dependencies in go.mod
- Removed unused functions from fileutil package
- Removed unused hash package

### Build Process
- Updated Makefile to accept MAVEN_INDEX_DIR parameter
- Modified build command to use maven-index directory structure
- Removed db-crawl target from Makefile

### GitHub Actions
- Updated workflow to clone maven-index repository
- Removed crawl step from build process
- Added support for forks using github.repository_owner
- Updated to use ORG_REPO_TOKEN for private repo access

### Documentation
- Updated README.md to reflect new data source

## Usage

Build database from maven-index:
```bash
make db-build MAVEN_INDEX_DIR=../maven-index
```

The database building process reads TSV files from the maven-index
repository and creates the SQLite database as before.

knqyf263 committed:
ci: comment out upload step for testing

Temporarily disabled the upload assets step to verify the build process
works correctly with the new maven-index integration.

knqyf263 commented:
It works, although I didn't push DBs.
https://github.com/aquasecurity/trivy-java-db/actions/runs/16197437045/job/45727506764

knqyf263 committed:
Revert "ci: comment out upload step for testing"

This reverts commit 19c4864f443c448ed95422ff11f1387f5cad5f8a.

knqyf263 committed:
refactor: always create new database instead of updating

- Removed db.Exists() function as it's no longer needed
- Simplified build() function to always create a new database
- Removed conditional logic for existing database updates
- Updated Makefile to conditionally use --index-dir only when MAVEN_INDEX_DIR is specified

knqyf263 committed:
refactor: use errors.Is for EOF checking

Use errors.Is(err, io.EOF) instead of direct comparison for more robust
error handling, especially when errors are wrapped.

knqyf263 committed:
ci: always use aquasecurity/maven-index in workflow

Add comment explaining why we don't use github.repository_owner:
- maven-index is a private repository that cannot be forked
- Using dynamic owner would fail in forks due to lack of access

knqyf263 commented:
The testing database is available here.
https://github.com/knqyf263/trivy-java-db/pkgs/container/trivy-java-db

I don't see differences except for the database sizes.

```shell
$ trivy image tomcat:9.0.1 --scanners vuln --skip-version-check --format json -o tomcat.json --list-all-pkgs
2025-07-11T11:21:53+04:00	INFO	[vuln] Vulnerability scanning is enabled
2025-07-11T11:21:56+04:00	INFO	[javadb] Downloading Java DB...
2025-07-11T11:21:56+04:00	INFO	[javadb] Downloading artifact...	repo="mirror.gcr.io/aquasec/trivy-java-db:1"
745.65 MiB / 745.65 MiB [---------------------------------------------------------------------------------------] 100.00% 20.46 MiB p/s 37s
2025-07-11T11:22:35+04:00	INFO	[javadb] Artifact successfully downloaded	repo="mirror.gcr.io/aquasec/trivy-java-db:1"
2025-07-11T11:22:35+04:00	INFO	[javadb] Java DB is cached for 3 days. If you want to update the database more frequently, "trivy clean --java-db" command clears the DB cache.
2025-07-11T11:22:35+04:00	INFO	Detected OS	family="debian" version="9.2"
2025-07-11T11:22:35+04:00	INFO	[debian] Detecting vulnerabilities...	os_version="9" pkg_num=287
2025-07-11T11:22:35+04:00	INFO	Number of language-specific files	num=1
2025-07-11T11:22:35+04:00	INFO	[jar] Detecting vulnerabilities...
2025-07-11T11:22:35+04:00	WARN	Using severities from other vendors for some vulnerabilities. Read https://trivy.dev/dev/docs/scanner/vulnerability#severity-selection for details.
2025-07-11T11:22:35+04:00	WARN	This OS version is no longer supported by the distribution	family="debian" version="9.2"
2025-07-11T11:22:35+04:00	WARN	The vulnerability detection may be insufficient because security updates are not provided

$ jq '.Results[].Packages | length' tomcat.json
287
29

$ trivy clean --java-db --scan-cache
2025-07-11T11:21:36+04:00	INFO	Removing scan cache...
2025-07-11T11:21:36+04:00	INFO	Removing Java database...

$ trivy image tomcat:9.0.1 --java-db-repository ghcr.io/knqyf263/trivy-java-db:1 --scanners vuln --skip-version-check --format json -o knqyf263-tomcat.json
2025-07-11T11:18:31+04:00	INFO	[vuln] Vulnerability scanning is enabled
2025-07-11T11:18:42+04:00	INFO	[javadb] Downloading Java DB...
2025-07-11T11:18:42+04:00	INFO	[javadb] Downloading artifact...	repo="ghcr.io/knqyf263/trivy-java-db:1"
754.16 MiB / 754.16 MiB [---------------------------------------------------------------------------------------] 100.00% 16.99 MiB p/s 45s
2025-07-11T11:19:28+04:00	INFO	[javadb] Artifact successfully downloaded	repo="ghcr.io/knqyf263/trivy-java-db:1"
2025-07-11T11:19:28+04:00	INFO	[javadb] Java DB is cached for 3 days. If you want to update the database more frequently, "trivy clean --java-db" command clears the DB cache.
2025-07-11T11:19:28+04:00	INFO	Detected OS	family="debian" version="9.2"
2025-07-11T11:19:28+04:00	INFO	[debian] Detecting vulnerabilities...	os_version="9" pkg_num=287
2025-07-11T11:19:28+04:00	INFO	Number of language-specific files	num=1
2025-07-11T11:19:28+04:00	INFO	[jar] Detecting vulnerabilities...
2025-07-11T11:19:29+04:00	WARN	Using severities from other vendors for some vulnerabilities. Read https://trivy.dev/dev/docs/scanner/vulnerability#severity-selection for details.
2025-07-11T11:19:29+04:00	WARN	This OS version is no longer supported by the distribution	family="debian" version="9.2"
2025-07-11T11:19:29+04:00	WARN	The vulnerability detection may be insufficient because security updates are not provided
2025-07-11T11:19:29+04:00	INFO	Table result includes only package filenames. Use '--format json' option to get the full path to the package file.

$ jq '.Results[].Packages | length' knqyf263-tomcat.json
287
29
```

knqyf263 commented:
@DmitriyLewen I believe it's ready for review now.

DmitriyLewen commented on cmd/trivy-java-db/main.go:
```diff
@@ -57,44 +48,31 @@ func init() {
-			return xerrors.Errorf("db init error: %w", err)
-		}
+	
+	if err = dbc.Init(); err != nil {
```
We get error here (if db/trivy-java.db exists):
```
Error: db init error: unable to create 'artifacts' table: SQL logic error: table artifacts already exists (1)
```

knqyf263 replied:
I added a protection so that the existing database will not be deleted by mistake.
https://github.com/aquasecurity/trivy-java-db/pull/60/commits/90a2ec3b5a7ca85380d799c411b7a7269f0759a8

DmitriyLewen commented on pkg/builder/builder.go:
```diff
@@ -33,49 +35,89 @@ func NewBuilder(db db.DB, meta db.Client) Builder {
-func (b *Builder) Build(cacheDir string) error {
-	indexDir := filepath.Join(cacheDir, "indexes")
+func (b *Builder) Build(indexDir string) error {
+	slog.Info("Building the index database")
```
nit:
```suggestion
	slog.Info("Building the index database", slog.String("index dir", indexDir))
```

knqyf263 replied:
Thanks. Added https://github.com/aquasecurity/trivy-java-db/pull/60/commits/23361b384b185e98a2e66a611592a24fdb162bcb

DmitriyLewen commented on pkg/types/types.go:
It looks like we can remove IndexDir

knqyf263 replied:
Removed in https://github.com/aquasecurity/trivy-java-db/pull/60/commits/0ae836c1f3d2492ee6f184d50d5dc663ef1573f8

DmitriyLewen commented on pkg/index/reader.go:
```diff
@@ -0,0 +1,47 @@
+	csv.Reader
+}
+
+func Open(path string) (*Reader, error) {
```
This is unused function. Do you plan to use it?
if no - we can remove `f *os.File` and `Clsoe()` function.

knqyf263 replied:
Yes, I plan to import it from maven-index-crawler since they should share the TSV reader.

- maven-index-crawler: Write TSV files (and read them for caching)
- trivy-java-db: Read TSV files

knqyf263 replied:
Added a comment: https://github.com/aquasecurity/trivy-java-db/pull/60/commits/b2d6433170875c82e599cd8c7abf3d1f154d192a

DmitriyLewen reviewed (commented):
left comments.

I also imported this into Trivy.
- We didn't remove required structures.
- Tomcat image scan completed without errors.

knqyf263 committed:
feat: prevent overwriting existing database

- Add check for existing database before building
- Return error with clear message suggesting 'make clean'
- Re-add db.Exists() function for database existence check

knqyf263 committed:
feat: make cache directory configurable in Makefile

- Add CACHE_DIR variable with default value ./cache
- Use CACHE_DIR in db-build, db-compress, and clean targets
- Allow users to specify custom cache directory: make db-build CACHE_DIR=./custom-cache

knqyf263 committed:
feat: add index directory to build log

Add index_dir to the build log message to improve debugging
and visibility of which directory is being processed.

knqyf263 committed:
refactor: remove unused IndexesDir constant

Remove IndexesDir constant from types.go as it's no longer used
after the crawler removal.

knqyf263 committed:
docs: add documentation for Open function

Add comprehensive documentation for the Open function explaining:
- Purpose: reading Maven artifact index data from TSV files
- Usage: used by external repositories to read compiled database
- Format: tab-separated values with 5 fields (GroupID, ArtifactID, Version, Classifier, SHA1)

DmitriyLewen commented on .github/workflows/cron.yml:
```diff
@@ -29,33 +29,33 @@ jobs:
-          username: ${{ secrets.DOCKERHUB_USER }}
-          password: ${{ secrets.DOCKERHUB_TOKEN }}
+      # Commented out for fork testing - Docker Hub and ECR credentials not available
+      # - name: Login to Docker Hub
```
Don't forget to revert that commented out code back in this file.
And also bellow.

DmitriyLewen reviewed (approved):
LGTM

knqyf263 committed:
fix: improve comment formatting in workflow

Fix comment line break in maven-index repository explanation.

knqyf263 committed:
fix: restore Exists function implementation and position

- Use os.IsNotExist() for more explicit error checking
- Move Exists function back to line 30 to minimize diff
- Maintain original function order

knqyf263 commented:
Triggered https://github.com/aquasecurity/trivy-java-db/actions/runs/16261303046

knqyf263 commented:
Seems fine. The database size: 745.65 MiB => 754.18 MiB

```
$ trivy image tomcat:9.0.1 --scanners vuln --skip-version-check --java-db-repository ghcr.io/aquasecurity/trivy-java-db:1 --format json -o tomcat.json --list-all-pkgs
2025-07-14T12:45:02+04:00	INFO	[vuln] Vulnerability scanning is enabled
2025-07-14T12:45:06+04:00	INFO	[javadb] Downloading Java DB...
2025-07-14T12:45:06+04:00	INFO	[javadb] Downloading artifact...	repo="ghcr.io/aquasecurity/trivy-java-db:1"
754.18 MiB / 754.18 MiB [---------------------------------------------------------------------------------------] 100.00% 17.90 MiB p/s 42s
2025-07-14T12:45:49+04:00	INFO	[javadb] Artifact successfully downloaded	repo="ghcr.io/aquasecurity/trivy-java-db:1"
2025-07-14T12:45:49+04:00	INFO	[javadb] Java DB is cached for 3 days. If you want to update the database more frequently, "trivy clean --java-db" command clears the DB cache.
2025-07-14T12:45:49+04:00	INFO	Detected OS	family="debian" version="9.2"
2025-07-14T12:45:49+04:00	INFO	[debian] Detecting vulnerabilities...	os_version="9" pkg_num=287
2025-07-14T12:45:49+04:00	INFO	Number of language-specific files	num=1
2025-07-14T12:45:49+04:00	INFO	[jar] Detecting vulnerabilities...
2025-07-14T12:45:49+04:00	WARN	Using severities from other vendors for some vulnerabilities. Read https://trivy.dev/dev/docs/scanner/vulnerability#severity-selection for details.
2025-07-14T12:45:49+04:00	WARN	This OS version is no longer supported by the distribution	family="debian" version="9.2"
2025-07-14T12:45:49+04:00	WARN	The vulnerability detection may be insufficient because security updates are not provided

$ jq '.Results[].Packages | length' tomcat.json
287
29
```

knqyf263 commented:
@DmitriyLewen Please let me know if you see any problems.
